File: src/systems/Pathfinding.ts
/**
 * 4-directional A* on a tile grid.
 *
 * The grid is represented as a flat array of "blocked" booleans for speed.
 * We expose a high-level `findRoute` that walks the ordered waypoint list and
 * concatenates segments — this is what the build phase consults to decide
 * whether a placement disconnects the creep route.
 */

import { Cell, GRID_H, GRID_W, WAYPOINTS, Waypoint } from '../data/map';

export interface Point {
  x: number;
  y: number;
}

export type Blocked = (x: number, y: number) => boolean;

const DIRS = [
  [1, 0],
  [-1, 0],
  [0, 1],
  [0, -1],
] as const;

/**
 * Min-heap keyed by f-score. Small custom impl avoids dep weight.
 * Stores [f, idx] pairs; idx is a packed (y*GRID_W + x) tile index.
 */
class MinHeap {
  private heap: { f: number; idx: number }[] = [];
  push(f: number, idx: number): void {
    this.heap.push({ f, idx });
    this.bubbleUp(this.heap.length - 1);
  }
  pop(): { f: number; idx: number } | undefined {
    if (this.heap.length === 0) return undefined;
    const top = this.heap[0];
    const last = this.heap.pop()!;
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.sinkDown(0);
    }
    return top;
  }
  get size(): number {
    return this.heap.length;
  }
  private bubbleUp(i: number): void {
    while (i > 0) {
      const parent = (i - 1) >> 1;
      if (this.heap[parent].f <= this.heap[i].f) break;
      [this.heap[parent], this.heap[i]] = [this.heap[i], this.heap[parent]];
      i = parent;
    }
  }
  private sinkDown(i: number): void {
    const n = this.heap.length;
    while (true) {
      const l = 2 * i + 1;
      const r = 2 * i + 2;
      let smallest = i;
      if (l < n && this.heap[l].f < this.heap[smallest].f) smallest = l;
      if (r < n && this.heap[r].f < this.heap[smallest].f) smallest = r;
      if (smallest === i) break;
      [this.heap[smallest], this.heap[i]] = [this.heap[i], this.heap[smallest]];
      i = smallest;
    }
  }
}

function manhattan(ax: number, ay: number, bx: number, by: number): number {
  return Math.abs(ax - bx) + Math.abs(ay - by);
}

/**
 * A* between two grid tiles.
 * @returns ordered list of points (inclusive of start and goal) or null if unreachable.
 */
export function aStar(
  start: Point,
  goal: Point,
  blocked: Blocked,
  bounds: { w: number; h: number } = { w: GRID_W, h: GRID_H },
): Point[] | null {
  const w = bounds.w;
  const h = bounds.h;
  if (start.x === goal.x && start.y === goal.y) return [{ ...start }];

  const startIdx = start.y * w + start.x;
  const goalIdx = goal.y * w + goal.x;

  const cameFrom = new Map<number, number>();
  const gScore = new Map<number, number>();
  const open = new MinHeap();

  gScore.set(startIdx, 0);
  open.push(manhattan(start.x, start.y, goal.x, goal.y), startIdx);

  while (open.size > 0) {
    const top = open.pop()!;
    const idx = top.idx;
    if (idx === goalIdx) {
      return reconstruct(cameFrom, idx, w);
    }
    const cx = idx % w;
    const cy = (idx - cx) / w;
    const cg = gScore.get(idx)!;
    for (const [dx, dy] of DIRS) {
      const nx = cx + dx;
      const ny = cy + dy;
      if (nx < 0 || ny < 0 || nx >= w || ny >= h) continue;
      // The goal tile itself may be "blocked" (e.g. a Path tile considered
      // walkable for creeps but flagged as path) — leave that to the caller's
      // `blocked` predicate.
      if (blocked(nx, ny)) continue;
      const nIdx = ny * w + nx;
      const tentative = cg + 1;
      const prev = gScore.get(nIdx);
      if (prev === undefined || tentative < prev) {
        cameFrom.set(nIdx, idx);
        gScore.set(nIdx, tentative);
        const f = tentative + manhattan(nx, ny, goal.x, goal.y);
        open.push(f, nIdx);
      }
    }
  }
  return null;
}

function reconstruct(cameFrom: Map<number, number>, endIdx: number, w: number): Point[] {
  const out: Point[] = [];
  let curr: number | undefined = endIdx;
  while (curr !== undefined) {
    const x = curr % w;
    const y = (curr - x) / w;
    out.push({ x, y });
    curr = cameFrom.get(curr);
  }
  return out.reverse();
}

/**
 * Build a `blocked` predicate from a grid + an optional set of
 * tentative-blocked tiles (used to test whether placing a tower
 * would disconnect the path).
 */
export function blockedFromGrid(grid: Cell[][], extra?: Set<string>): Blocked {
  return (x, y) => {
    if (extra?.has(`${x},${y}`)) return true;
    const c = grid[y][x];
    return c === Cell.Wall || c === Cell.Tower || c === Cell.Rock;
  };
}

/** Find a route through all waypoints; null if any segment is unreachable. */
export function findRoute(
  grid: Cell[][],
  extra?: Set<string>,
  waypoints: readonly Waypoint[] = WAYPOINTS,
): Point[][] | null {
  const blocked = blockedFromGrid(grid, extra);
  const segments: Point[][] = [];
  for (let i = 0; i < waypoints.length - 1; i++) {
    const a = waypoints[i];
    const b = waypoints[i + 1];
    const seg = aStar(a, b, blocked);
    if (!seg) return null;
    segments.push(seg);
  }
  return segments;
}

/** Concatenate segments into a single ordered point list. */
export function flattenRoute(segments: Point[][]): Point[] {
  const out: Point[] = [];
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];
    if (i === 0) out.push(...seg);
    else out.push(...seg.slice(1));
  }
  return out;
}

/** Straight-line route through waypoints for air creeps (Bresenham-style). */
export function buildAirRoute(waypoints: readonly Waypoint[] = WAYPOINTS): Point[] {
  const out: Point[] = [];
  for (let i = 0; i < waypoints.length - 1; i++) {
    const a = waypoints[i];
    const b = waypoints[i + 1];
    const steps = Math.max(Math.abs(b.x - a.x), Math.abs(b.y - a.y));
    for (let s = (i === 0 ? 0 : 1); s <= steps; s++) {
      const t = steps === 0 ? 0 : s / steps;
      out.push({
        x: Math.round(a.x + (b.x - a.x) * t),
        y: Math.round(a.y + (b.y - a.y) * t),
      });
    }
  }
  return out;
}

File: tools/maze_optimizer/pathfinding.py
import heapq

import numpy as np

from grid import GRID_W, GRID_H, Cell, WAYPOINTS, FOOTPRINT

DIRS = ((1, 0), (-1, 0), (0, 1), (0, -1))


def _manhattan(ax: int, ay: int, bx: int, by: int) -> int:
    return abs(ax - bx) + abs(ay - by)


def a_star(
    start: tuple[int, int],
    goal: tuple[int, int],
    grid: np.ndarray,
    extra: set[tuple[int, int]] | None = None,
) -> list[tuple[int, int]] | None:
    sx, sy = start
    gx, gy = goal
    if sx == gx and sy == gy:
        return [(sx, sy)]

    goal_idx = gy * GRID_W + gx
    start_idx = sy * GRID_W + sx

    g_score: dict[int, int] = {start_idx: 0}
    came_from: dict[int, int] = {}

    counter = 0
    heap: list[tuple[int, int, int]] = []
    heapq.heappush(heap, (_manhattan(sx, sy, gx, gy), counter, start_idx))

    has_extra = extra is not None and len(extra) > 0

    while heap:
        f, _, idx = heapq.heappop(heap)
        if idx == goal_idx:
            return _reconstruct(came_from, idx)

        cx = idx % GRID_W
        cy = idx // GRID_W
        cg = g_score.get(idx)
        if cg is None or f - _manhattan(cx, cy, gx, gy) > cg:
            continue

        for dx, dy in DIRS:
            nx = cx + dx
            ny = cy + dy
            if nx < 0 or ny < 0 or nx >= GRID_W or ny >= GRID_H:
                continue
            if has_extra and (nx, ny) in extra:
                continue
            c = grid[ny, nx]
            if c == Cell.Wall or c == Cell.Tower or c == Cell.Rock:
                continue
            n_idx = ny * GRID_W + nx
            tentative = cg + 1
            prev = g_score.get(n_idx)
            if prev is not None and tentative >= prev:
                continue
            came_from[n_idx] = idx
            g_score[n_idx] = tentative
            counter += 1
            heapq.heappush(
                heap,
                (tentative + _manhattan(nx, ny, gx, gy), counter, n_idx),
            )

    return None


def _reconstruct(came_from: dict[int, int], end_idx: int) -> list[tuple[int, int]]:
    path: list[tuple[int, int]] = []
    curr: int | None = end_idx
    while curr is not None:
        path.append((curr % GRID_W, curr // GRID_W))
        curr = came_from.get(curr)
    path.reverse()
    return path


def find_route(
    grid: np.ndarray,
    extra: set[tuple[int, int]] | None = None,
    waypoints=None,
) -> list[list[tuple[int, int]]] | None:
    if waypoints is None:
        waypoints = WAYPOINTS
    segments: list[list[tuple[int, int]]] = []
    for i in range(len(waypoints) - 1):
        a = waypoints[i]
        b = waypoints[i + 1]
        seg = a_star((a.x, a.y), (b.x, b.y), grid, extra)
        if seg is None:
            return None
        segments.append(seg)
    return segments


def flatten_route(segments: list[list[tuple[int, int]]]) -> list[tuple[int, int]]:
    out: list[tuple[int, int]] = []
    for i, seg in enumerate(segments):
        if i == 0:
            out.extend(seg)
        else:
            out.extend(seg[1:])
    return out


def footprint_cells(x: int, y: int) -> set[tuple[int, int]]:
    return {(x + dx, y + dy) for dx, dy in FOOTPRINT}


try:
    from pathfinding_cy import a_star as _fast_a_star
except ImportError:
    _fast_a_star = a_star


def build_cell_to_seg(
    segments: list[list[tuple[int, int]]],
) -> dict[tuple[int, int], set[int]]:
    mapping: dict[tuple[int, int], set[int]] = {}
    for seg_idx, seg in enumerate(segments):
        for cell in seg:
            if cell in mapping:
                mapping[cell].add(seg_idx)
            else:
                mapping[cell] = {seg_idx}
    return mapping


def reroute_affected(
    grid: np.ndarray,
    segments: list[list[tuple[int, int]]],
    affected_cells: set[tuple[int, int]],
    cell_to_seg: dict[tuple[int, int], set[int]],
    extra: set[tuple[int, int]] | None = None,
    waypoints=None,
) -> list[list[tuple[int, int]]] | None:
    if waypoints is None:
        waypoints = WAYPOINTS
    dirty: set[int] = set()
    for cell in affected_cells:
        s = cell_to_seg.get(cell)
        if s:
            dirty.update(s)
    if not dirty:
        return segments
    new_segments = list(segments)
    for seg_idx in dirty:
        a = waypoints[seg_idx]
        b = waypoints[seg_idx + 1]
        new_seg = _fast_a_star((a[0], a[1]), (b[0], b[1]), grid, extra)
        if new_seg is None:
            return None
        new_segments[seg_idx] = new_seg
    return new_segments
